### ISSUE TITLE
Broken Pad: TypeError: r.dropdowns is undefined - Changeset related. - Waiting on review.

### ISSUE THREAD

**Issue by JohnMcLear** [ether/etherpad-lite#3861]
I finally have a clear bug that causes the r.dropdowns.

1. Minify is on w/ high maxage
1. Database is MySQL
1. Instance is medium traffic/volume but nothing concerning (loads super low).
1. Instance is runing in production mode.

## Steps to replicate
When user visits Etherpad the pad doesn't fully render.

## Things I tested.
1. .etherpad file can be exported from /export/etherpad
1. checkPad.js and extradPadData.js are no use.
1. Restart changes nothing.
1. /export/txt returns a blank pad with no warnings or errors in console.

## Error in console:
```
Error: mismatch error setting raw text in setDocAText testBrokenPad:798:13
    setDocAText https://hidden.com/p/testBrokenPad line 24 > scriptElement line 232 > Function:798
    setDocumentAttributedText https://hidden.com/p/testBrokenPad line 24 > scriptElement line 232 > Function:206
    setBaseAttributedText https://hidden.com/p/testBrokenPad line 24 > scriptElement line 232 > Function:10300
    withCallbacks https://hidden.com/p/testBrokenPad line 24 > scriptElement line 232 > Function:202
    inCallStack https://hidden.com/p/testBrokenPad line 24 > scriptElement line 232 > Function:492
    inCallStackIfNecessary https://hidden.com/p/testBrokenPad line 24 > scriptElement line 232 > Function:561
    withCallbacks https://hidden.com/p/testBrokenPad line 24 > scriptElement line 232 > Function:199
    setBaseAttributedText https://hidden.com/p/testBrokenPad line 24 > scriptElement line 232 > Function:10286
    ace_setBaseAttributedText https://hidden.com/p/testBrokenPad line 24 > scriptElement line 232 > Function:1017
    Ace2Editor/</editor[fnName]< https://hidden.com/javascripts/lib/ep_etherpad-lite/static/js/pad.js?callback=require.define:2857
    action https://hidden.com/javascripts/lib/ep_etherpad-lite/static/js/pad.js?callback=require.define:2810
    doActionsPendingInit https://hidden.com/javascripts/lib/ep_etherpad-lite/static/js/pad.js?callback=require.define:2830
    forEach https://hidden.com/javascripts/lib/ep_etherpad-lite/static/js/ace2_common.js?callback=require.define:12709
    doActionsPendingInit https://hidden.com/javascripts/lib/ep_etherpad-lite/static/js/pad.js?callback=require.define:2829
    onEditorReady https://hidden.com/javascripts/lib/ep_etherpad-lite/static/js/pad.js?callback=require.define:2962
    readyFunc https://hidden.com/p/testBrokenPad line 3097 > scriptElement:17
    init https://hidden.com/p/testBrokenPad line 24 > scriptElement line 232 > Function:5510
```

## Further error can be produced by hitting /export/txt

```
Error: Failed assertion: mismatched composition of two changesets testBrokenPad:5971:11
    error https://hidden.com/p/testBrokenPad/export/etherpad line 24 > scriptElement line 232 > Function:5971
    assert https://hidden.com/p/testBrokenPad/export/etherpad line 24 > scriptElement line 232 > Function:5985
    compose https://hidden.com/p/testBrokenPad/export/etherpad line 24 > scriptElement line 232 > Function:7253
    composeUserChangeset https://hidden.com/p/testBrokenPad/export/etherpad line 24 > scriptElement line 232 > Function:10313
    doRepApplyChangeset https://hidden.com/p/testBrokenPad/export/etherpad line 24 > scriptElement line 232 > Function:2254
    performDocumentApplyChangeset https://hidden.com/p/testBrokenPad/export/etherpad line 24 > scriptElement line 232 > Function:2077
    performDocumentReplaceRange https://hidden.com/p/testBrokenPad/export/etherpad line 24 > scriptElement line 232 > Function:2323
    performDocumentReplaceSelection https://hidden.com/p/testBrokenPad/export/etherpad line 24 > scriptElement line 232 > Function:2489
    doReturnKey https://hidden.com/p/testBrokenPad/export/etherpad line 24 > scriptElement line 232 > Function:3440
    handleKeyEvent https://hidden.com/p/testBrokenPad/export/etherpad line 24 > scriptElement line 232 > Function:3867
    inCallStack https://hidden.com/p/testBrokenPad/export/etherpad line 24 > scriptElement line 232 > Function:492
    inCallStackIfNecessary https://hidden.com/p/testBrokenPad/export/etherpad line 24 > scriptElement line 232 > Function:561
    handleKeyEvent https://hidden.com/p/testBrokenPad/export/etherpad line 24 > scriptElement line 232 > Function:3716
    dispatch https://hidden.com/p/testBrokenPad/export/etherpad line 24 > scriptElement line 232 > Function:5633
    handle https://hidden.com/p/testBrokenPad/export/etherpad line 24 > scriptElement line 232 > Function:5285
```

## Exported pad.etherpad with beautified json.
```
{
  "pad:testBrokenPad":{
    "atext":{
      "text":"",
      "attribs":""
      
    },
    "pool":{
      "numToAttrib":{
        "0":[
          "author",
          "a.Vm7xRPrAz5SPOrMe"
          
        ]
        
      },
      "nextNum":1
      
    },
    "head":83,
    "chatHead":5,
    "publicStatus":false,
    "passwordHash":null,
    "savedRevisions":[
      
      
    ]
    
  },
  "globalAuthor:a.Vm7xRPrAz5SPOrMe":{
    "colorId":"#c3cdee",
    "name":"Noura",
    "timestamp":1586750805643,
    "padIDs":"testBrokenPad"
    
  },
  "pad:testBrokenPad:revs:0":{
    "changeset":"Z:1>1eo|c+191+5n$WELCOME TO YOUR ONLINE CONSULTATION! Below is a brief overview of this module. For more information, click the question mark at the top right of the screen.\n\nWHITEBOARD: This area of the screen is the document collaboration whiteboard. Here, you can import or paste a document, or type text. Changes made to text in this window are highlighted and seen immediately by both individuals participating in the online consultation.\n\nTEXT CHAT: You can use the area to the right of the screen to have a text conversation in real time. \n\nTOOLBAR: The icons on the left side of the toolbar allow you to work with a document's formatting. The icons on the right side (or at the bottom if using a phone) include options for your online session, such as importing or exporting a document. Hover over any icon for a text label showing the icon's function.\n\nDRAWING TOOLS: Clicking the pencil icon allows you to draw on top of the whiteboard. Within the drawing area, you can change the brush, clear your drawing, or use a solid background. You can also import images and export the entire drawing area. On a computer, hover over the drawing area to expand it, and hover away from it to minimize it. On a touch screen, touch the pencil icon once to open the drawing function, twice to expand the drawing area, and a third time to close it.\n\nAUDIO AND VIDEO: If your center has enabled audio and video consultations, you can click on the camera icon to allow the other participant to see and hear you. The first time you use audio and video, your browser may prompt you to allow the use of your camera and microphone in your session.\n\nAFTER YOUR SESSION: Your chat transcript and document will be saved in this online meeting. You can always come back by viewing your appointment and clicking the 'start or join online consultation' link.",
    "meta":{
      "author":"",
      "timestamp":1586750416483,
      "atext":{
        "text":"WELCOME TO YOUR ONLINE CONSULTATION! Below is a brief overview of this module. For more information, click the question mark at the top right of the screen.\n\nWHITEBOARD: This area of the screen is the document collaboration whiteboard. Here, you can import or paste a document, or type text. Changes made to text in this window are highlighted and seen immediately by both individuals participating in the online consultation.\n\nTEXT CHAT: You can use the area to the right of the screen to have a text conversation in real time. \n\nTOOLBAR: The icons on the left side of the toolbar allow you to work with a document's formatting. The icons on the right side (or at the bottom if using a phone) include options for your online session, such as importing or exporting a document. Hover over any icon for a text label showing the icon's function.\n\nDRAWING TOOLS: Clicking the pencil icon allows you to draw on top of the whiteboard. Within the drawing area, you can change the brush, clear your drawing, or use a solid background. You can also import images and export the entire drawing area. On a computer, hover over the drawing area to expand it, and hover away from it to minimize it. On a touch screen, touch the pencil icon once to open the drawing function, twice to expand the drawing area, and a third time to close it.\n\nAUDIO AND VIDEO: If your center has enabled audio and video consultations, you can click on the camera icon to allow the other participant to see and hear you. The first time you use audio and video, your browser may prompt you to allow the use of your camera and microphone in your session.\n\nAFTER YOUR SESSION: Your chat transcript and document will be saved in this online meeting. You can always come back by viewing your appointment and clicking the 'start or join online consultation' link.\n",
        "attribs":"|d+1ep"
        
      }
      
    }    
  }
}
```

**Comment by JohnMcLear** [comment]
Tasks to do
 - [ ] Discover which changeset is causing the pad to die.
 - [ ] Discover why that changeset is being accepted
 - [ ] Create better client/server side handling should a document get broken like this.

This is a very old / rare bug so it's nice to have debug capability but all in all there can always be times when a pad goes into a broken state, data is never 100% accurate, we should always have a method for checking pad state prior to shipping it to a user.  For now I'm going to focus on removing a lot of the json to find out the offending changesets.  I should get time to work on this in a few days.

**Comment by JohnMcLear** [comment]
```
  "pad:testBrokenPad":{
    "atext":{
      "text":"",
      "attribs":""
      
    },
```
is the problem..

Changing it to 
```
  "pad:testBrokenPad":{
    "atext":{
      "text":"\n",
      "attribs":""
      
    },
```

Solves the problem..


Task:
 - [ ] Discover how did atext become completely null?
 - [ ] Enforce atext is never stored as ``""`` but if it's ``""`` it gets stored as "\n"

Working full blob for curious..

```
{
  "pad:testBrokenPad":{
    "atext":{
      "text":"\n",
      "attribs":""

    },
    "pool":{
      "numToAttrib":{
        "0":[
          "author",
          "a.Vm7xRPrAz5SPOrMe"

        ]

      },
      "nextNum":1

    },
    "head":1,
    "chatHead":-1,
    "publicStatus":false,
    "passwordHash":null,
    "savedRevisions":[


    ]

  },
  "globalAuthor:a.Vm7xRPrAz5SPOrMe":{
    "colorId":"#c3cdee",
    "name":"Noura",
    "timestamp":1586750805643,
    "padIDs":"testBrokenPad"

  },
  "pad:testBrokenPad:revs:0":{
    "changeset":"Z:1>1eo|c+191+5n$WELCOME TO YOUR ONLINE CONSULTATION! Below is a brief overview of this module. For more information, click the question mark at the top right of the screen.\n\nWHITEBOARD: This area of the screen is the document collaboration whiteboard. Here, you can import or paste a document, or type text. Changes made to text in this window are highlighted and seen immediately by both individuals participating in the online consultation.\n\nTEXT CHAT: You can use the area to the right of the screen to have a text conversation in real time. \n\nTOOLBAR: The icons on the left side of the toolbar allow you to work with a document's formatting. The icons on the right side (or at the bottom if using a phone) include options for your online session, such as importing or exporting a document. Hover over any icon for a text label showing the icon's function.\n\nDRAWING TOOLS: Clicking the pencil icon allows you to draw on top of the whiteboard. Within the drawing area, you can change the brush, clear your drawing, or use a solid background. You can also import images and export the entire drawing area. On a computer, hover over the drawing area to expand it, and hover away from it to minimize it. On a touch screen, touch the pencil icon once to open the drawing function, twice to expand the drawing area, and a third time to close it.\n\nAUDIO AND VIDEO: If your center has enabled audio and video consultations, you can click on the camera icon to allow the other participant to see and hear you. The first time you use audio and video, your browser may prompt you to allow the use of your camera and microphone in your session.\n\nAFTER YOUR SESSION: Your chat transcript and document will be saved in this online meeting. You can always come back by viewing your appointment and clicking the 'start or join online consultation' link.",
    "meta":{
      "author":"",
      "timestamp":1586750416483,
      "atext":{
        "text":"WELCOME TO YOUR ONLINE CONSULTATION! Below is a brief overview of this module. For more information, click the question mark at the top right of the screen.\n\nWHITEBOARD: This area of the screen is the document collaboration whiteboard. Here, you can import or paste a document, or type text. Changes made to text in this window are highlighted and seen immediately by both individuals participating in the online consultation.\n\nTEXT CHAT: You can use the area to the right of the screen to have a text conversation in real time. \n\nTOOLBAR: The icons on the left side of the toolbar allow you to work with a document's formatting. The icons on the right side (or at the bottom if using a phone) include options for your online session, such as importing or exporting a document. Hover over any icon for a text label showing the icon's function.\n\nDRAWING TOOLS: Clicking the pencil icon allows you to draw on top of the whiteboard. Within the drawing area, you can change the brush, clear your drawing, or use a solid background. You can also import images and export the entire drawing area. On a computer, hover over the drawing area to expand it, and hover away from it to minimize it. On a touch screen, touch the pencil icon once to open the drawing function, twice to expand the drawing area, and a third time to close it.\n\nAUDIO AND VIDEO: If your center has enabled audio and video consultations, you can click on the camera icon to allow the other participant to see and hear you. The first time you use audio and video, your browser may prompt you to allow the use of your camera and microphone in your session.\n\nAFTER YOUR SESSION: Your chat transcript and document will be saved in this online meeting. You can always come back by viewing your appointment and clicking the 'start or join online consultation' link.\n",
        "attribs":"|d+1ep"

      }

    }

  },
  "pad:testBrokenPad:revs:1":{
    "changeset":"Z:1ep<1|2=4e=38-1$",
    "meta":{
      "author":"a.Vm7xRPrAz5SPOrMe",
      "timestamp":1586750431380

    }

  }
}
```

**Comment by JohnMcLear** [comment]
RE export not erroring, it's probably it wasn't erroring..  The server doesn't have a problem with completly empty contents.  It's only the browser that does..  My client side fix #3863 will do the job here and imho should be merged w/ 1.8.3 but I'd recommend further investigation to HOW atext of "" is even workable..  Like can a ``setText`` operation from the APi set pad contents to ``""`` or does it always get delivered as ``"\n"```? etc.

Anyway, anyone can pick this up now and we have a patch that gives stability, it can be moved down the priority list once my patch is merged.

**Comment by JohnMcLear** [comment]
FWIW this problem is what was causing the tests to fail too @lpagliari did a clearPad operation which set pad contents to "" causing everything post that operation to fail :|

**Comment by muxator** [comment]
> I finally have a clear bug that causes the r.dropdowns.

@JohnMcLear, can we link similar issues here for future reference?

**Comment by JohnMcLear** [comment]
Sure :)

**Comment by muxator** [comment]
#3863 is merged. Let's add some references to similar issues from the past and close this.

**Comment by muxator** [comment]
I am not finding anything related to `r.dropdowns` searching in the issues.
Do we have something, @JohnMcLear?

Otherwise please close.

**Comment by JohnMcLear** [comment]
It can be closed.  Data is here for posterity.
